reference-implementation/lib/readable-stream.js

```diff
@@ -156,67 +131,349 @@ export default class ReadableStream {
+    return ReleaseReadableStreamReader(this);
   }
+}
 
-  _initReadyPromise() {
-    this._readyPromise = new Promise((resolve, reject) => {
-      this._readyPromise_resolve = resolve;
-    });
+function AcquireReadableStreamReader(stream) {
+  return new ReadableStreamReader(stream);
+}
+
+function CallReadableStreamPull(stream) {
+  if (stream._draining === true || stream._started === false ||
+      stream._state === 'closed' || stream._state === 'errored' ||
+      stream._pullScheduled === true) {
+    return undefined;
   }
 
-  _initClosedPromise() {
-    this._closedPromise = new Promise((resolve, reject) => {
-      this._closedPromise_resolve = resolve;
-      this._closedPromise_reject = reject;
+  if (stream._pullingPromise !== undefined) {
+    stream._pullScheduled = true;
+    stream._pullingPromise.then(() => {
+      stream._pullScheduled = false;
+      CallReadableStreamPull(stream);
     });
+    return undefined;
+  }
+
+  const shouldApplyBackpressure = ShouldReadableStreamApplyBackpressure(stream);
+  if (shouldApplyBackpressure === true) {
+    return undefined;
   }
 
-  // Note: The resolve function and reject function are cleared when the
-  // corresponding promise is resolved or rejected. This is for debugging. This
-  // makes extra resolve/reject calls for the same promise fail so that we can
-  // detect unexpected extra resolve/reject calls that may be caused by bugs in
-  // the algorithm.
+  stream._pullingPromise = PromiseInvokeOrNoop(stream._underlyingSource, 'pull', [stream._enqueue, stream._close]);
+  stream._pullingPromise.then(
+    () => { stream._pullingPromise = undefined; },
+    e => { stream._error(e); }
+  );
 
-  _resolveReadyPromise(value) {
-    this._readyPromise_resolve(value);
-    this._readyPromise_resolve = null;
+  return undefined;
+}
+
+function CancelReadableStream(stream, reason) {
+  if (stream._state === 'closed') {
+    return Promise.resolve(undefined);
+  }
+  if (stream._state === 'errored') {
+    return Promise.reject(stream._storedError);
   }
 
-  _resolveClosedPromise(value) {
-    this._closedPromise_resolve(value);
-    this._closedPromise_resolve = null;
-    this._closedPromise_reject = null;
+  stream._queue = [];
+  CloseReadableStream(stream);
+
+  const sourceCancelPromise = PromiseInvokeOrNoop(stream._underlyingSource, 'cancel', [reason]);
+  return sourceCancelPromise.then(() => undefined);
+}
+
+function CloseReadableStream(stream) {
+  assert(stream._state === 'readable');
+
+  stream._state = 'closed';
+
+  if (IsReadableStreamLocked(stream) === true) {
+    return ReleaseReadableStreamReader(stream._reader);
   }
 
-  _rejectClosedPromise(reason) {
-    this._closedPromise_reject(reason);
-    this._closedPromise_resolve = null;
-    this._closedPromise_reject = null;
+  return undefined;
+}
+
+function CreateReadableStreamCloseFunction(stream) {
+  return () => {
+    if (stream._state !== 'readable') {
+      return undefined;
```

Comment:
Hmm. I think you are right. Underlying source should be more careful than this. Open a new issue?